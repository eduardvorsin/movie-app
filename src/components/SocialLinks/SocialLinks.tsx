import { GeneralProps, SocialNetworksItem } from '@/types/shared';
import Link from '@/components/Link/Link';
import { twMerge } from 'tailwind-merge';

const socialNetworkIcons = {
	imdb: '/assets/icons/imdb.svg#imdb',
	twitter: '/assets/icons/twitter.svg#twitter',
	youtube: '/assets/icons/youtube.svg#youtube',
	tiktok: '/assets/icons/tiktok.svg#tiktok',
	facebook: '/assets/icons/facebook.svg#facebook',
	instagram: '/assets/icons/instagram.svg#instagram',
} as const;

type Props = {
	data: SocialNetworksItem[],
} & GeneralProps;

export default function SocialLinks({
	className,
	testId,
	data,
	...props
}: Props) {
	const classes = twMerge(
		'flex items-center flex-wrap gap-2',
		className
	);

	return (
		<div
			className={classes}
			{...props}
		>
			{data.map((item) => (
				<Link
					key={item.name}
					data-testid={testId}
					className='text-0 flex w-10 h-10 fill-neutral-300 text-dark-neutral-300 hover:fill-neutral-400 hover:text-dark-neutral-250 active:fill-dark-neutral-500 active:text-dark-neutral-200 dark:fill-dark-neutral-400 dark:text-neutral-300 dark:hover:fill-dark-neutral-500 dark:hover:text-neutral-200 dark:active:fill-dark-neutral-600 dark:active:text-neutral-100 transition-colors duration-150'
					href={item.url}
					isExternal
					target='_blank'
				>
					<svg viewBox='0 0 32 32'>
						<use href={socialNetworkIcons[item.name]}></use>
					</svg>
					<span className='sr-only'>
						{`${item.name[0].toUpperCase()}${item.name.slice(1)}`}
					</span>
				</Link>
			))}
		</div>
	);
}