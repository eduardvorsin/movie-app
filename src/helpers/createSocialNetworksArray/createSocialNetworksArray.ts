import { ExternalIDS } from '@/services/types';
import { KeyWithoutId, SocialNetworksItem } from '@/types/shared';

const socialNetworkUrls = {
	imdb_id: 'https://www.imdb.com/',
	facebook_id: 'https://www.facebook.com/',
	instagram_id: 'https://www.instagram.com/',
	tiktok_id: 'https://www.tiktok.com/@',
	twitter_id: 'https://twitter.com/',
	youtube_id: 'https://www.youtube.com/@',
} as const;

const socialNetworkFields = new Set<keyof ExternalIDS>([
	'imdb_id',
	'facebook_id',
	'instagram_id',
	'tiktok_id',
	'twitter_id',
	'youtube_id',
]);

export const createSocialNetworksArray = (data: ExternalIDS): SocialNetworksItem[] => {
	const keys = Object.keys(data) as Array<keyof ExternalIDS>;

	return keys.filter((key) => socialNetworkFields.has(key) && data[key]).map((key) => ({
		name: key.slice(0, -3) as KeyWithoutId<typeof key>,
		url: `${socialNetworkUrls[key]}${data[key]}`,
	}));
};