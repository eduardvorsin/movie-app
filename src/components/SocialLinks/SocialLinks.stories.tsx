import type { Meta, StoryObj } from '@storybook/react';
import SocialLinks from './SocialLinks';
import { createSocialNetworksArray } from '@/helpers/createSocialNetworksArray/createSocialNetworksArray';

const meta: Meta<typeof SocialLinks> = {
	title: 'UI/SocialLinks',
	tags: ['autodocs'],
	argTypes: {
		className: {
			description: 'Additional classes for the component'
		},
		testId: {
			description: 'Id for testing the component',
		},
		isRounded: {
			description: 'Allows you to make icons round',
		},
		data: {
			control: false,
			description: 'An array for generating links to social networks',
		},
	},
	component: SocialLinks,
	parameters: {
		docs: {
			description: {
				component: 'SocialLinks is an interface element designed to display and provide access to various social networks and platforms. This component is often used to integrate and provide users with easy access to social profiles or pages associated with a web application.',
			},
		},
	},
}

export default meta;
type Story = StoryObj<typeof SocialLinks>;

const data = createSocialNetworksArray({
	imdb_id: null,
	facebook_id: ' ',
	instagram_id: ' ',
	tiktok_id: ' ',
	twitter_id: ' ',
	youtube_id: ' ',
});

export const Default: Story = {
	args: { data },
	parameters: {
		docs: {
			description: {
				story: 'The default state of the component',
			},
		},
	},
};

export const Rounded: Story = {
	args: { data, isRounded: true },
	parameters: {
		docs: {
			description: {
				story: 'Makes social media icons round',
			},
		},
	},
};