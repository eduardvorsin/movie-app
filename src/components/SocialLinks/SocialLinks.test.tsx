import { render, screen } from "@testing-library/react";
import SocialLinks from "./SocialLinks";
import { createSocialNetworksArray } from "@/helpers/createSocialNetworksArray/createSocialNetworksArray";

const mockData = createSocialNetworksArray({
	imdb_id: null,
	facebook_id: ' ',
	instagram_id: ' ',
	tiktok_id: ' ',
	twitter_id: ' ',
	youtube_id: ' ',
});

describe('SocialLinks tests', () => {
	it('is rendered correctly', () => {
		render(
			<SocialLinks
				data={mockData}
				testId='test-social-links'
			/>
		);

		expect(screen.getByTestId<HTMLDivElement>('test-social-links')).toBeInTheDocument();
	});

	it('is a basic snapshot', () => {
		render(
			<SocialLinks
				data={mockData}
				testId='test-social-links'
			/>
		);

		expect(screen.getByTestId<HTMLDivElement>('test-social-links')).toMatchSnapshot();
	});

	it('a snapshot with isRounded set to true', () => {
		render(
			<SocialLinks
				data={mockData}
				isRounded
				testId='test-social-links'
			/>
		);

		expect(screen.getByTestId<HTMLDivElement>('test-social-links')).toMatchSnapshot();
	});
});