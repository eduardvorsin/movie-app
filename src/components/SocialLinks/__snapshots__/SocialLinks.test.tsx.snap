// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SocialLinks tests a snapshot with isRounded set to true 1`] = `
<div
  class="flex items-center flex-wrap gap-2 "
  data-testid="test-social-links"
>
  <a
    aria-disabled="false"
    class="no-underline transition-colors duration-150 text-blue-800 hover:text-blue-900 active:text-blue-1000 dark:text-blue-400 dark:hover:text-blue-300 dark:active:text-blue-200 hover:underline  text-0 flex w-10 h-10 fill-neutral-300 [&]:text-dark-neutral-300 hover:fill-neutral-400 hover:[&]:text-dark-neutral-250 active:fill-dark-neutral-500 active:[&]:text-dark-neutral-200 dark:fill-dark-neutral-300 dark:[&]:text-neutral-300 dark:hover:fill-dark-neutral-400 dark:hover:[&]:text-neutral-200 dark:active:fill-dark-neutral-500 dark:active:[&]:text-neutral-100"
    href="https://www.facebook.com/ "
    rel="noopener noreferrer"
    target="_blank"
  >
    <svg
      class="rounded-full"
      viewBox="0 0 32 32"
    >
      <use
        href="/assets/icons/facebook.svg#facebook"
      />
    </svg>
    <span
      class="sr-only"
    >
      Facebook
    </span>
  </a>
  <a
    aria-disabled="false"
    class="no-underline transition-colors duration-150 text-blue-800 hover:text-blue-900 active:text-blue-1000 dark:text-blue-400 dark:hover:text-blue-300 dark:active:text-blue-200 hover:underline  text-0 flex w-10 h-10 fill-neutral-300 [&]:text-dark-neutral-300 hover:fill-neutral-400 hover:[&]:text-dark-neutral-250 active:fill-dark-neutral-500 active:[&]:text-dark-neutral-200 dark:fill-dark-neutral-300 dark:[&]:text-neutral-300 dark:hover:fill-dark-neutral-400 dark:hover:[&]:text-neutral-200 dark:active:fill-dark-neutral-500 dark:active:[&]:text-neutral-100"
    href="https://www.instagram.com/ "
    rel="noopener noreferrer"
    target="_blank"
  >
    <svg
      class="rounded-full"
      viewBox="0 0 32 32"
    >
      <use
        href="/assets/icons/instagram.svg#instagram"
      />
    </svg>
    <span
      class="sr-only"
    >
      Instagram
    </span>
  </a>
  <a
    aria-disabled="false"
    class="no-underline transition-colors duration-150 text-blue-800 hover:text-blue-900 active:text-blue-1000 dark:text-blue-400 dark:hover:text-blue-300 dark:active:text-blue-200 hover:underline  text-0 flex w-10 h-10 fill-neutral-300 [&]:text-dark-neutral-300 hover:fill-neutral-400 hover:[&]:text-dark-neutral-250 active:fill-dark-neutral-500 active:[&]:text-dark-neutral-200 dark:fill-dark-neutral-300 dark:[&]:text-neutral-300 dark:hover:fill-dark-neutral-400 dark:hover:[&]:text-neutral-200 dark:active:fill-dark-neutral-500 dark:active:[&]:text-neutral-100"
    href="https://www.tiktok.com/@ "
    rel="noopener noreferrer"
    target="_blank"
  >
    <svg
      class="rounded-full"
      viewBox="0 0 32 32"
    >
      <use
        href="/assets/icons/tiktok.svg#tiktok"
      />
    </svg>
    <span
      class="sr-only"
    >
      Tiktok
    </span>
  </a>
  <a
    aria-disabled="false"
    class="no-underline transition-colors duration-150 text-blue-800 hover:text-blue-900 active:text-blue-1000 dark:text-blue-400 dark:hover:text-blue-300 dark:active:text-blue-200 hover:underline  text-0 flex w-10 h-10 fill-neutral-300 [&]:text-dark-neutral-300 hover:fill-neutral-400 hover:[&]:text-dark-neutral-250 active:fill-dark-neutral-500 active:[&]:text-dark-neutral-200 dark:fill-dark-neutral-300 dark:[&]:text-neutral-300 dark:hover:fill-dark-neutral-400 dark:hover:[&]:text-neutral-200 dark:active:fill-dark-neutral-500 dark:active:[&]:text-neutral-100"
    href="https://twitter.com/ "
    rel="noopener noreferrer"
    target="_blank"
  >
    <svg
      class="rounded-full"
      viewBox="0 0 32 32"
    >
      <use
        href="/assets/icons/twitter.svg#twitter"
      />
    </svg>
    <span
      class="sr-only"
    >
      Twitter
    </span>
  </a>
  <a
    aria-disabled="false"
    class="no-underline transition-colors duration-150 text-blue-800 hover:text-blue-900 active:text-blue-1000 dark:text-blue-400 dark:hover:text-blue-300 dark:active:text-blue-200 hover:underline  text-0 flex w-10 h-10 fill-neutral-300 [&]:text-dark-neutral-300 hover:fill-neutral-400 hover:[&]:text-dark-neutral-250 active:fill-dark-neutral-500 active:[&]:text-dark-neutral-200 dark:fill-dark-neutral-300 dark:[&]:text-neutral-300 dark:hover:fill-dark-neutral-400 dark:hover:[&]:text-neutral-200 dark:active:fill-dark-neutral-500 dark:active:[&]:text-neutral-100"
    href="https://www.youtube.com/@ "
    rel="noopener noreferrer"
    target="_blank"
  >
    <svg
      class="rounded-full"
      viewBox="0 0 32 32"
    >
      <use
        href="/assets/icons/youtube.svg#youtube"
      />
    </svg>
    <span
      class="sr-only"
    >
      Youtube
    </span>
  </a>
</div>
`;

exports[`SocialLinks tests is a basic snapshot 1`] = `
<div
  class="flex items-center flex-wrap gap-2 "
  data-testid="test-social-links"
>
  <a
    aria-disabled="false"
    class="no-underline transition-colors duration-150 text-blue-800 hover:text-blue-900 active:text-blue-1000 dark:text-blue-400 dark:hover:text-blue-300 dark:active:text-blue-200 hover:underline  text-0 flex w-10 h-10 fill-neutral-300 [&]:text-dark-neutral-300 hover:fill-neutral-400 hover:[&]:text-dark-neutral-250 active:fill-dark-neutral-500 active:[&]:text-dark-neutral-200 dark:fill-dark-neutral-300 dark:[&]:text-neutral-300 dark:hover:fill-dark-neutral-400 dark:hover:[&]:text-neutral-200 dark:active:fill-dark-neutral-500 dark:active:[&]:text-neutral-100"
    href="https://www.facebook.com/ "
    rel="noopener noreferrer"
    target="_blank"
  >
    <svg
      class=""
      viewBox="0 0 32 32"
    >
      <use
        href="/assets/icons/facebook.svg#facebook"
      />
    </svg>
    <span
      class="sr-only"
    >
      Facebook
    </span>
  </a>
  <a
    aria-disabled="false"
    class="no-underline transition-colors duration-150 text-blue-800 hover:text-blue-900 active:text-blue-1000 dark:text-blue-400 dark:hover:text-blue-300 dark:active:text-blue-200 hover:underline  text-0 flex w-10 h-10 fill-neutral-300 [&]:text-dark-neutral-300 hover:fill-neutral-400 hover:[&]:text-dark-neutral-250 active:fill-dark-neutral-500 active:[&]:text-dark-neutral-200 dark:fill-dark-neutral-300 dark:[&]:text-neutral-300 dark:hover:fill-dark-neutral-400 dark:hover:[&]:text-neutral-200 dark:active:fill-dark-neutral-500 dark:active:[&]:text-neutral-100"
    href="https://www.instagram.com/ "
    rel="noopener noreferrer"
    target="_blank"
  >
    <svg
      class=""
      viewBox="0 0 32 32"
    >
      <use
        href="/assets/icons/instagram.svg#instagram"
      />
    </svg>
    <span
      class="sr-only"
    >
      Instagram
    </span>
  </a>
  <a
    aria-disabled="false"
    class="no-underline transition-colors duration-150 text-blue-800 hover:text-blue-900 active:text-blue-1000 dark:text-blue-400 dark:hover:text-blue-300 dark:active:text-blue-200 hover:underline  text-0 flex w-10 h-10 fill-neutral-300 [&]:text-dark-neutral-300 hover:fill-neutral-400 hover:[&]:text-dark-neutral-250 active:fill-dark-neutral-500 active:[&]:text-dark-neutral-200 dark:fill-dark-neutral-300 dark:[&]:text-neutral-300 dark:hover:fill-dark-neutral-400 dark:hover:[&]:text-neutral-200 dark:active:fill-dark-neutral-500 dark:active:[&]:text-neutral-100"
    href="https://www.tiktok.com/@ "
    rel="noopener noreferrer"
    target="_blank"
  >
    <svg
      class=""
      viewBox="0 0 32 32"
    >
      <use
        href="/assets/icons/tiktok.svg#tiktok"
      />
    </svg>
    <span
      class="sr-only"
    >
      Tiktok
    </span>
  </a>
  <a
    aria-disabled="false"
    class="no-underline transition-colors duration-150 text-blue-800 hover:text-blue-900 active:text-blue-1000 dark:text-blue-400 dark:hover:text-blue-300 dark:active:text-blue-200 hover:underline  text-0 flex w-10 h-10 fill-neutral-300 [&]:text-dark-neutral-300 hover:fill-neutral-400 hover:[&]:text-dark-neutral-250 active:fill-dark-neutral-500 active:[&]:text-dark-neutral-200 dark:fill-dark-neutral-300 dark:[&]:text-neutral-300 dark:hover:fill-dark-neutral-400 dark:hover:[&]:text-neutral-200 dark:active:fill-dark-neutral-500 dark:active:[&]:text-neutral-100"
    href="https://twitter.com/ "
    rel="noopener noreferrer"
    target="_blank"
  >
    <svg
      class=""
      viewBox="0 0 32 32"
    >
      <use
        href="/assets/icons/twitter.svg#twitter"
      />
    </svg>
    <span
      class="sr-only"
    >
      Twitter
    </span>
  </a>
  <a
    aria-disabled="false"
    class="no-underline transition-colors duration-150 text-blue-800 hover:text-blue-900 active:text-blue-1000 dark:text-blue-400 dark:hover:text-blue-300 dark:active:text-blue-200 hover:underline  text-0 flex w-10 h-10 fill-neutral-300 [&]:text-dark-neutral-300 hover:fill-neutral-400 hover:[&]:text-dark-neutral-250 active:fill-dark-neutral-500 active:[&]:text-dark-neutral-200 dark:fill-dark-neutral-300 dark:[&]:text-neutral-300 dark:hover:fill-dark-neutral-400 dark:hover:[&]:text-neutral-200 dark:active:fill-dark-neutral-500 dark:active:[&]:text-neutral-100"
    href="https://www.youtube.com/@ "
    rel="noopener noreferrer"
    target="_blank"
  >
    <svg
      class=""
      viewBox="0 0 32 32"
    >
      <use
        href="/assets/icons/youtube.svg#youtube"
      />
    </svg>
    <span
      class="sr-only"
    >
      Youtube
    </span>
  </a>
</div>
`;
